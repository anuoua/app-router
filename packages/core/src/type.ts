export interface BasicTreeNode<R = unknown> {
  name: string;
  route: string | undefined;
  path: string;
  parent?: BasicTreeNode<R>;
  children: BasicTreeNode<R>[];
  resolved?: R;
}
