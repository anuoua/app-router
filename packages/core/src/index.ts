export * from "./utils/build";
export * from "./utils/match";
export * from "./utils/sort";
export * from "./type";
