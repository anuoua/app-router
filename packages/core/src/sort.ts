import type { BasicTreeNode } from "./type";
import {
  isExactName,
  isExactSlug,
  isGroup,
  isOptionalRestSlug,
  isPrivate,
  isRestSlug,
} from "./utils";

const WEIGHTS = {
  Exact: 9,
  ExactSlug: 8,
  RestSlug: 7,
  OptinalRestSlug: 6,
};

const MAX_SEGMENTS = 10;

export const getScore = (route: string) => {
  const segments = route.split("/");

  segments.shift();

  let score = 0;

  for (let i = 0; i < segments.length; i++) {
    const segment = segments[i];
    const levelWeights = 10 ** (MAX_SEGMENTS - i);
    if (isExactName(segment)) {
      score += levelWeights * WEIGHTS.Exact;
      continue;
    }
    if (isOptionalRestSlug(segment)) {
      score += levelWeights * WEIGHTS.OptinalRestSlug;
      continue;
    }
    if (isRestSlug(segment)) {
      score += levelWeights * WEIGHTS.RestSlug;
      continue;
    }
    if (isExactSlug(segment)) {
      score += levelWeights * WEIGHTS.ExactSlug;
      continue;
    }
  }
  return score;
};

export const sort = <T extends BasicTreeNode>(
  nodes: [T, Record<string, any>][]
) => {
  return [...nodes].sort(([a], [b]) => getScore(b.route!) - getScore(a.route!));
};
