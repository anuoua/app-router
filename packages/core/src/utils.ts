export const privateReg = /^\_\w+$/;
export const parallelReg = /^\@\w+$/;
export const interceptionReg = /^\(\./;
export const groupReg = /^\(\w+\)$/;
export const exactNameReg = /^\w+$/;
export const exactSlugReg = /^\[(\w+)\]$/;
export const restSlugReg = /^\[\.\.\.(\w+)\]$/;
export const optionalRestSlugReg = /^\[\[\.\.\.(\w+)\]\]$/;

export const isPrivate = (segment: string) => privateReg.test(segment);

export const isParallel = (segment: string) => parallelReg.test(segment);

export const isInterception = (segment: string) =>
  interceptionReg.test(segment);

export const isGroup = (segment: string) => groupReg.test(segment);

export const isExactName = (segment: string) => exactNameReg.test(segment);

export const isExactSlug = (segment: string) => exactSlugReg.test(segment);

export const isRestSlug = (segment: string) => restSlugReg.test(segment);

export const isOptionalRestSlug = (segment: string) =>
  optionalRestSlugReg.test(segment);

export const mergeParams = (params: Record<string, any>[]) =>
  Object.assign({}, ...params);
