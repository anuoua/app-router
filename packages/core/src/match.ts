import { sort } from "./sort";
import type { BasicTreeNode } from "./type";
import {
  exactSlugReg,
  isExactName,
  isExactSlug,
  isGroup,
  isOptionalRestSlug,
  isRestSlug,
  mergeParams,
  optionalRestSlugReg,
  restSlugReg,
} from "./utils";

export const match = <T>(path: string, tree: BasicTreeNode<T>[]) => {
  const segments = path.split("/");
  segments.shift();

  const matched: [BasicTreeNode<T>, Record<string, any>][] = [];
  const paramsStack: Record<string, any>[] = [];

  const walk = (segmentIndex: number, children: BasicTreeNode<T>[]) => {
    const segment = segments[segmentIndex];

    for (let i = 0; i < children.length; i++) {
      const node = children[i];

      if (isGroup(node.name)) {
        walk(segmentIndex, node.children);
      } else if (isOptionalRestSlug(node.name)) {
        const match = node.name.match(optionalRestSlugReg) as RegExpMatchArray;
        paramsStack.push({
          [match[1]]: segments.slice(segmentIndex),
        });
        matched.push([node, mergeParams([...paramsStack])]);
        paramsStack.pop();
      } else if (isRestSlug(node.name)) {
        const match = node.name.match(restSlugReg) as RegExpMatchArray;
        paramsStack.push({
          [match[1]]: segments.slice(segmentIndex),
        });
        matched.push([node, mergeParams([...paramsStack])]);
        paramsStack.pop();
      } else if (isExactName(node.name)) {
        if (node.name === segment) {
          const isNextSegmentExist = !!segments[segmentIndex + 1];
          if (isNextSegmentExist) {
            walk(segmentIndex + 1, node.children);
          } else {
            const lastOptionalNode = node.children.find((child) =>
              isOptionalRestSlug(child.name)
            );
            if (lastOptionalNode) {
              const match = lastOptionalNode.name.match(
                optionalRestSlugReg
              ) as RegExpMatchArray;
              paramsStack.push({
                [match[1]]: segments.slice(segmentIndex),
              });
              matched.push([lastOptionalNode, mergeParams([...paramsStack])]);
              paramsStack.pop();
            } else {
              matched.push([node, mergeParams([...paramsStack])]);
            }
          }
        }
      } else if (isExactSlug(node.name)) {
        const isNextSegmentExist = !!segments[segmentIndex + 1];
        const lastOptionalNode = node.children.find((child) =>
          isOptionalRestSlug(child.name)
        );
        const match = node.name.match(exactSlugReg) as RegExpMatchArray;
        paramsStack.push({
          [match[1]]: segment,
        });
        if (!isNextSegmentExist && !lastOptionalNode) {
          matched.push([node, mergeParams([...paramsStack])]);
        } else {
          walk(segmentIndex + 1, node.children);
        }
        paramsStack.pop();
      } else {
        throw new Error(`Route name: "${node.name}" is not supported`);
      }
    }
  };

  const [root] = tree;

  if (segments[0] === "") {
    matched.push([root, []]);
  } else {
    walk(0, root.children);
  }

  const [first] = sort(matched);

  return first as typeof first | undefined;
};
