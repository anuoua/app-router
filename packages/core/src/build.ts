import path from "path";
import { glob } from "glob";
import { isGroup, isParallel, isInterception, isPrivate } from "./utils";
import type { BasicTreeNode } from "./type";

const buildTree = async (
  routePaths: string[],
  target: string,
  options: Options
) => {
  const root: BasicTreeNode = {
    name: "",
    route: "",
    path: "",
    children: [],
  };

  const routeMap: Record<string, BasicTreeNode> = {};

  for (let routePath of routePaths) {
    const segments = (routePath === "." ? "" : `/${routePath}`).split("/");
    let currentNode: BasicTreeNode = root;

    for (let i = 0; i < segments.length; i++) {
      const segment = segments[i];

      if (isPrivate(segment) || isParallel(segment) || isInterception(segment))
        break;

      let childNode = currentNode.children?.find(
        (child) => child.name === segment
      );

      if (!childNode) {
        let route =
          segments
            .filter((i) => !isGroup(i))
            .slice(0, i + 1)
            .join("/") || "/";
        childNode = {
          name: segment === "/" ? "" : segment,
          route: isGroup(segment) ? undefined : route,
          path: path.resolve(target, routePath),
          children: [],
        };
        childNode.parent = currentNode === root ? undefined : currentNode;
        childNode.resolved = await options.resolve?.(childNode);
        currentNode.children?.push(childNode);
        if (!isGroup(segment)) routeMap[route] = childNode;
      }

      currentNode = childNode;
    }
  }

  return root.children || [];
};

export interface Options {
  resolve?: (node: BasicTreeNode) => Promise<any>;
}

export const build = async <P extends Options>(target: string, options: P) => {
  const allPaths = await glob("./**/", {
    cwd: path.resolve(target),
  });
  type Resolve = (typeof options)["resolve"];
  type RetType = Resolve extends (...args: any[]) => Promise<infer R>
    ? R
    : undefined;
  return buildTree(allPaths, target, options) as unknown as Promise<
    BasicTreeNode<RetType>[]
  >;
};
