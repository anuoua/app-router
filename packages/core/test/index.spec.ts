import { it, expect } from "bun:test";
import { build } from "../src/build";

it("read dir", async () => {
  const result = await build("test/app", {
    resolve: async (node) => {
      // @ts-expect-error
      delete node.path;
    },
  });
  expect(result).toMatchSnapshot();
});
