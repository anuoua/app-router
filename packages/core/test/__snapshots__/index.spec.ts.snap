// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`read dir 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [],
            "name": "post",
            "parent": [Circular],
            "resolved": undefined,
            "route": "/post",
          },
          {
            "children": [],
            "name": "draft",
            "parent": [Circular],
            "resolved": undefined,
            "route": "/draft",
          },
        ],
        "name": "(group)",
        "parent": [Circular],
        "resolved": undefined,
        "route": undefined,
      },
      {
        "children": [
          {
            "children": [],
            "name": "[id]",
            "parent": [Circular],
            "resolved": undefined,
            "route": "/home/[id]",
          },
          {
            "children": [],
            "name": "[[...optional]]",
            "parent": [Circular],
            "resolved": undefined,
            "route": "/home/[[...optional]]",
          },
          {
            "children": [],
            "name": "park",
            "parent": [Circular],
            "resolved": undefined,
            "route": "/home/park",
          },
        ],
        "name": "home",
        "parent": [Circular],
        "resolved": undefined,
        "route": "/home",
      },
      {
        "children": [
          {
            "children": [
              {
                "children": [],
                "name": "[...rest]",
                "parent": [Circular],
                "resolved": undefined,
                "route": "/settings/[user]/[...rest]",
              },
              {
                "children": [],
                "name": "admin",
                "parent": [Circular],
                "resolved": undefined,
                "route": "/settings/[user]/admin",
              },
              {
                "children": [
                  {
                    "children": [],
                    "name": "inner",
                    "parent": [Circular],
                    "resolved": undefined,
                    "route": "/settings/[user]/[role]/inner",
                  },
                ],
                "name": "[role]",
                "parent": [Circular],
                "resolved": undefined,
                "route": "/settings/[user]/[role]",
              },
            ],
            "name": "[user]",
            "parent": [Circular],
            "resolved": undefined,
            "route": "/settings/[user]",
          },
        ],
        "name": "settings",
        "parent": [Circular],
        "resolved": undefined,
        "route": "/settings",
      },
    ],
    "name": "",
    "parent": undefined,
    "resolved": undefined,
    "route": "/",
  },
]
`;
